src/auto-evo/EvolutionaryTree.cs

```diff
@@ -0,0 +1,222 @@
+
+    private PackedScene treeNodeScene = null!;
+
+    private Vector2 treeNodeSize;
+
+    private uint maxSpeciesId;
+
+    private int latestGeneration;
+
+    [Signal]
+    public delegate void SpeciesSelected(int generation, uint id);
+
+    public override void _Ready()
+    {
+        base._Ready();
+
+        treeNodeScene = GD.Load<PackedScene>("res://src/auto-evo/EvolutionaryTreeNode.tscn");
+
+        var tempNode = treeNodeScene.Instance<EvolutionaryTreeNode>();
+        treeNodeSize = tempNode.RectSize;
+        tempNode.QueueFree();
+
+        latoSmall = GD.Load<Font>("res://src/gui_common/fonts/Lato-Italic-Small.tres");
+    }
+
+    public override void _Draw()
+    {
+        base._Draw();
+
+        // Draw timeline
+        DrawLine(new Vector2(0, TIMELINE_LINE_Y), new Vector2(RectSize.x, TIMELINE_LINE_Y), Colors.DarkCyan,
+            TIMELINE_LINE_THICKNESS, true);
+
+        for (int i = 0; i <= latestGeneration; i++)
+        {
+            DrawLine(new Vector2(i * GENERATION_SEPARATION + treeNodeSize.x / 2, TIMELINE_LINE_Y),
+                new Vector2(i * GENERATION_SEPARATION + treeNodeSize.x / 2, TIMELINE_LINE_Y + TIMELINE_MARK_SIZE),
+                Colors.DarkCyan, TIMELINE_LINE_THICKNESS, true);
+
+            var localizedText = i + " " + TranslationServer.Translate("MEGA_YEARS");
+            var size = latoSmall.GetStringSize(localizedText);
+            DrawString(latoSmall, new Vector2(i * GENERATION_SEPARATION + treeNodeSize.x / 2 - size.x / 2,
+                    TIMELINE_LINE_Y + TIMELINE_MARK_SIZE * 2 + size.y),
+                localizedText, Colors.DarkCyan);
+        }
+
+        // Draw node connection lines
+        foreach (var node in nodes)
+        {
+            if (node.ParentNode == null)
+                continue;
+
+            DrawLine(node.ParentNode.Center, node.Center);
+        }
+
+        // Draw lines that indicate the species goes on till current generation, and species name
+        foreach (var latestNode in latestNodes.Values.Where(n => !n.LastGeneration))
+        {
+            var lineStart = latestNode.Center;
+            var lineEnd = new Vector2(GENERATION_SEPARATION * latestGeneration + latestNode.RectSize.x, lineStart.y);
+            DrawLine(lineStart, lineEnd);
+            DrawString(latoSmall, lineEnd + new Vector2(SPECIES_NAME_OFFSET, 0), speciesNames[latestNode.SpeciesID]);
+        }
+
+        // Draw extinct species name
+        foreach (var extinctedSpecies in latestNodes.Values.Where(n => n.LastGeneration))
+        {
+            DrawString(latoSmall, new Vector2(
+                extinctedSpecies.RectPosition.x + extinctedSpecies.RectSize.x + SPECIES_NAME_OFFSET,
+                extinctedSpecies.Center.y), speciesNames[extinctedSpecies.SpeciesID], Colors.DarkRed);
+        }
+    }
+
+    public void Init(Species player)
+    {
+        SetupTreeNode(player, null, 0);
+
+        speciesOrigin.Add(player.ID, (uint.MaxValue, 0));
```

Comment:
Are the zeros in this method the generation or something? If so I'd check that the given `Species.Generation == 0` otherwise the logic might break at a surprising time.

Reply:
This generation has nothing to do with Species.Generation. This actually means "0 Myrs"